core/server/master/src/main/java/alluxio/master/file/meta/UfsSyncUtils.java

```diff
@@ -34,95 +33,110 @@ private UfsSyncUtils() {} // prevent instantiation
-    boolean ufsExists = !Constants.INVALID_UFS_FINGERPRINT.equals(ufsFingerprint);
-    boolean ufsIsDir = ufsFingerprint != null && Fingerprint.Type.DIRECTORY.name()
-        .equals(Fingerprint.parse(ufsFingerprint).getTag(Fingerprint.Tag.TYPE));
+    boolean isContentSynced = inodeUfsIsContentSynced(inode, ufsFingerprint);
+    boolean isMetadataSynced = inodeUfsIsMetadataSynced(inode, ufsFingerprint);
+    boolean ufsExists = ufsFingerprint.isValid();
+    boolean ufsIsDir = ufsFingerprint != null
+        && Fingerprint.Type.DIRECTORY.name().equals(ufsFingerprint.getTag(Fingerprint.Tag.TYPE));
 
     UfsSyncUtils.SyncPlan syncPlan = new UfsSyncUtils.SyncPlan();
-    if (!isSynced) {
-      // Alluxio inode is not synced with UFS, so update the inode metadata
-      // Updating an inode is achieved by deleting the inode, and then loading metadata.
-
-      if (inode.isDirectory() && (containsMountPoint || ufsIsDir)) {
-        // Instead of deleting and then loading metadata to update, try to update directly
-        // - mount points (or paths with mount point descendants) should not be deleted
-        // - directory permissions can be updated without removing the inode
-        if (inode.getParentId() != InodeTree.NO_PARENT) {
-          // Only update the inode if it is not the root directory. The root directory is a special
-          // case, since it is expected to be owned by the process that starts the master, and not
-          // the owner on UFS.
-          syncPlan.setUpdateDirectory();
-        }
-        syncPlan.setSyncChildren();
-      } else {
-        // update inode, by deleting and then optionally loading metadata
-        syncPlan.setDelete();
-        if (ufsExists) {
-          // UFS exists, so load metadata later.
-          syncPlan.setLoadMetadata();
-        }
-      }
-    } else {
+
+    if (isContentSynced && isMetadataSynced) {
       // Inode is already synced.
       if (inode.isDirectory() && inode.isPersisted()) {
         // Both Alluxio and UFS are directories, so sync the children of the directory.
         syncPlan.setSyncChildren();
       }
+      return syncPlan;
+    }
+
+    // One of the metadata or content is not in sync
+    if (inode.isDirectory() && (containsMountPoint || ufsIsDir)) {
+      // Instead of deleting and then loading metadata to update, try to update directly
+      // - mount points (or paths with mount point descendants) should not be deleted
+      // - directory permissions can be updated without removing the inode
+      if (inode.getParentId() != InodeTree.NO_PARENT) {
+        // Only update the inode if it is not the root directory. The root directory is a special
+        // case, since it is expected to be owned by the process that starts the master, and not
+        // the owner on UFS.
+        syncPlan.setUpdateMetadata();
+      }
+      syncPlan.setSyncChildren();
+      return syncPlan;
+    }
+
+    // One of metadata or content is not in sync and it is a file
```

Comment:
Can a directory reach this point too?

Reply:
The only way for a directory to reach this point, is when the inode says it is a directory and ufs says it is a file .  In which case, we do the delete and reload anyways.  I modified the comment to reflect that.